src/shared/sam/debugger/awsSamDebugger.ts

```diff
@@ -22,143 +32,335 @@ export const CODE_TARGET_TYPE: 'code' = 'code'
+
+        // XXX: return undefined, because we already launched and invoked by now.
+        //
+        // TODO: In the future we may consider NOT launching, and instead do one of the following:
+        //  - return a config here for vscode to handle
+        //  - return a config here for SamDebugSession.ts to handle (custom debug adapter)
         return undefined
     }
+
+    /**
+     * Launches a NodeJs lambda:
+     *
+     * 1. prepares a bunch of arguments
+     * 2. does `sam build`
+     * 3. does `sam local invoke`
+     *
+     * @param config  Launch-config generated by resolveDebugConfiguration()
+     * from a debug-config.
+     */
+    public async launchTypescript(config: SamLaunchRequestArgs) {
+        const launchConfig: NodejsDebugConfiguration = await tsDebug.makeTypescriptConfig(config)
+        await tsDebug.invokeTypescriptLambda(this.ctx, launchConfig)
+        return launchConfig
+    }
+
+    /**
+     * Launches a dotnet ("coreclr") lambda:
+     *
+     * 1. prepares a bunch of arguments
+     * 2. does `sam build`
+     * 3. does `sam local invoke`
+     *
+     * @param config  Launch-config generated by resolveDebugConfiguration()
+     * from a debug-config.
+     */
+    public async launchDotnet(config: SamLaunchRequestArgs) {
+        config = await csharpDebug.makeCsharpConfig(config)
+        await csharpDebug.invokeCsharpLambda(this.ctx, config)
+        return config
+    }
+
+    /**
+     * Launches a Python lambda:
+     *
+     * 1. prepares a bunch of arguments
+     * 2. does `sam build`
+     * 3. does `sam local invoke`
+     *
+     * @param config  Launch-config generated by resolveDebugConfiguration()
+     * from a debug-config.
+     */
+    public async launchPython(config: SamLaunchRequestArgs) {
+        //  Make a Python launch-config from the generic config.
+        const launchConfig: PythonDebugConfiguration = await pythonDebug.makePythonDebugConfig(
+            config,
+            !config.noDebug,
+            config.runtime,
+            config.handlerName,
+            config.documentUri,
+        )
+        await pythonDebug.invokePythonLambda(this.ctx, launchConfig)
+        return launchConfig
+    }
 }
 
-function createDirectInvokeSamDebugConfigurationFromTemplate(
-    resourceName: string,
-    templatePath: string
-): AwsSamDebuggerConfiguration {
-    return {
-        type: AWS_SAM_DEBUG_TYPE,
-        request: DIRECT_INVOKE_TYPE,
-        name: resourceName,
-        invokeTarget: {
-            target: TEMPLATE_TARGET_TYPE,
-            samTemplatePath: templatePath,
-            samTemplateResource: resourceName
+/**
+ * Validates debug configuration properties.
+ */
+function validateConfig(
+    folder: vscode.WorkspaceFolder | undefined,
+    config: AwsSamDebuggerConfiguration,
+): boolean {
+    const cftRegistry = CloudFormationTemplateRegistry.getRegistry()
+
+    let rv: { isValid: boolean; message?: string } = { isValid: false, message: undefined, }
```

Comment:
FYI: the only reason why we were using the object before is because the caller was in charge of outputting an error message. It's not needed if we're handling all messaging from within this function.

That being said, is this the right pattern? Should we leave it to the caller to provide the error message instead?

Reply:
fixed in https://github.com/aws/aws-toolkit-vscode/pull/1035